src/python/pants/backend/go/util_rules/build_pkg.py

```diff
@@ -22,26 +22,80 @@
-    direct_dependencies: tuple[BuildGoPackageRequest, ...]
+    def __init__(
+        self,
+        *,
+        import_path: str,
+        digest: Digest,
+        subpath: str,
+        go_file_names: tuple[str, ...],
+        s_file_names: tuple[str, ...],
+        direct_dependencies: tuple[BuildGoPackageRequest, ...],
+        minimum_go_version: str | None,
+        for_tests: bool = False,
+    ) -> None:
+        """Build a package and its dependencies as `__pkg__.a` files.
+
+        Instances of this class form a structure-shared DAG, and so a hashcode is pre-computed for
+        the recursive portion.
+        """
+
+        self.import_path = import_path
+        self.digest = digest
+        self.subpath = subpath
+        self.go_file_names = go_file_names
+        self.s_file_names = s_file_names
+        self.direct_dependencies = direct_dependencies
+        self.minimum_go_version = minimum_go_version
+        self.for_tests = for_tests
+        self._hashcode = hash(
+            (
+                self.import_path,
+                self.digest,
+                self.subpath,
+                self.go_file_names,
+                self.s_file_names,
+                self.direct_dependencies,
+                self.minimum_go_version,
+                self.for_tests,
+            )
+        )
 
-    # In the form `1.16`. This is declared in the `go` directive of `go.mod`.
-    minimum_go_version: str | None
+    def __repr__(self) -> str:
+        # NB: We must override the default `__repr__` so that `direct_dependencies` does not
+        # traverse into transitive dependencies, which was pathologically slow.
+        return (
+            f"{self.__class__}("
+            f"import_path={repr(self.import_path)}, "
+            f"digest={self.digest}, "
+            f"subpath={self.subpath}, "
+            f"go_file_names={self.go_file_names}, "
+            f"go_file_names={self.s_file_names}, "
+            f"direct_dependencies={[dep.import_path for dep in self.direct_dependencies]}, "
+            f"minimum_go_version={self.minimum_go_version}, "
+            f"for_tests={self.for_tests}"
+            ")"
+        )
 
-    for_tests: bool = False
+    def __hash__(self) -> int:
+        return self._hashcode
+
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            return NotImplemented
+        return (
+            self._hashcode == other._hashcode
+            and self.import_path == other.import_path
+            and self.digest == other.digest
+            and self.subpath == other.subpath
+            and self.go_file_names == other.go_file_names
+            and self.s_file_names == other.s_file_names
+            and self.minimum_go_version == other.minimum_go_version
+            and self.for_tests == other.for_tests
+            # TODO: Use a recursive memoized __eq__ if this ever shows up in profiles.
+            and self.direct_dependencies == other.direct_dependencies
```

Comment:
Maybe first compare the hash codes with `hash(self.direct_dependencies) == hash(other.direct_dependencies)` and then only check `self.direct_dependencies == other.direct_dependencies` if the hash codes were equal? Might help a bit with performance.

Reply:
We already check that `self._hashcode == other._hashcode`, so I doubt that buys us much. Generally, my concern is when they are `__eq__` how slow it takes to have to transitively walk `direct_dependencies`. First checking `hash()` would only make that situation worse.

Reply:
Ah I missed that.